File: src/boj/boj_18290.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_18290 {
    static final int[] dr = {0, 1, 0, -1};
    static final int[] dc = {1, 0, -1, 0};

    static int answer = Integer.MIN_VALUE;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        int[][] map = new int[n][m];
        for (int r = 0; r < n; r++) {
            st = new StringTokenizer(br.readLine());
            for (int c = 0; c < m; c++) {
                map[r][c] = Integer.parseInt(st.nextToken());
            }
        }

        recursive(n, m, k, 0, map, new boolean[n][m], 0);

        System.out.println(answer);
        br.close();
    }

    static void recursive(int n, int m, int k, int idx, int[][] map, boolean[][] visited, int cnt) {
        if (cnt == k) {
            int sum = 0;
            for (int r = 0; r < n; r++) {
                for (int c = 0; c < m; c++) {
                    if (visited[r][c]) {
                        sum += map[r][c];
                    }
                }
            }
            answer = Math.max(answer, sum);
            return;
        }

        outer: for (int i = idx; i < n * m; i++) {
            int r = i / m;
            int c = i % m;

            for (int d = 0; d < 4; d++) {
                int nr = r + dr[d];
                int nc = c + dc[d];
                if (0 <= nr && nr < n && 0 <= nc && nc < m && visited[nr][nc]) {
                    continue outer;
                }
            }

            visited[r][c] = true;
            recursive(n, m, k, i + 1, map, visited, cnt + 1);
            visited[r][c] = false;
        }
    }
}

File: src/boj/boj_18290.kt
package boj_18290

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max

var n = 0
var m = 0
var k = 0
lateinit var map: Array<IntArray>
var answer = Int.MIN_VALUE
val dx = arrayOf(1, -1, 0, 0)
val dy = arrayOf(0, 0, 1, -1)

fun main() {
    init()
    val visited = Array(n) { BooleanArray(m) }
    visited[0][0] = true
    backTracking(0, 0, visited, 1)
    visited[0][0] = false
    backTracking(0, 0, visited, 0)
    println(answer)
}

fun init() = with(BufferedReader(InputStreamReader(System.`in`))) {
    with(StringTokenizer(readLine())) {
        n = nextToken().toInt()
        m = nextToken().toInt()
        k = nextToken().toInt()
    }
    map = Array(n) {
        with(StringTokenizer(readLine())) {
            IntArray(m) {
                nextToken().toInt()
            }
        }
    }
    close()
}

fun backTracking(cx: Int, cy: Int, visited: Array<BooleanArray>, cnt: Int) {
    if (cnt == k) {
        var sum = 0
        for (y in 0 until n) {
            for (x in 0 until m) {
                if (visited[y][x]) {
                    sum += map[y][x]
                }
            }
        }
        answer = max(answer, sum)
        return
    }

    for (i in cx + cy * m + 1 until n * m) {
        val nx = i % m
        val ny = i / m
        if (canAdd(nx, ny, visited)) {
            visited[ny][nx] = true
            backTracking(nx, ny, visited, cnt + 1)
            visited[ny][nx] = false
        }
    }
}

fun canAdd(x: Int, y: Int, visited: Array<BooleanArray>): Boolean {
    for (d in 0 until 4) {
        val nx = x + dx[d]
        val ny = y + dy[d]
        if (nx in 0 until m && ny in 0 until n && visited[ny][nx]) {
            return false
        }
    }
    return true
}